Account/FxAPushMessageHandler.swift

```diff
@@ -0,0 +1,220 @@
+        }
+        return result
+    }
+}
+
+extension FxAPushMessageHandler {
+    func handleVerification() -> Success {
+        guard let account = profile.getAccount(), account.actionNeeded == .needsVerification else {
+            log.info("Account verified by server either doesn't exist or doesn't need verifying")
+            return deferMaybe(PushMessageError.noActionNeeded)
+        }
+
+        // Now we're verified, we can start syncing.
+        return account.advance().bind { _ in
+            return succeed()
+        }
+    }
+}
+
+/// An extension to handle each of the messages.
+extension FxAPushMessageHandler {
+    func handleDeviceConnected(_ data: JSON?) -> Success {
+        guard let deviceName = data?["deviceName"].string else {
+            return messageIncomplete(.deviceConnected)
+        }
+        return unimplemented(.deviceConnected, with: deviceName)
+    }
+}
+
+extension FxAPushMessageHandler {
+    func handleDeviceDisconnected(_ data: JSON?) -> Success {
+        guard let deviceID = data?["id"].string else {
+            return messageIncomplete(.deviceDisconnected)
+        }
+        return unimplemented(.deviceDisconnected, with: deviceID)
+    }
+}
+
+extension FxAPushMessageHandler {
+    func handleProfileUpdated() -> Success {
+        return unimplemented(.profileUpdated)
+    }
+}
+
+extension FxAPushMessageHandler {
+    func handlePasswordChanged() -> Success {
+        return unimplemented(.passwordChanged)
+    }
+}
+
+extension FxAPushMessageHandler {
+    func handlePasswordReset() -> Success {
+        return unimplemented(.passwordReset)
+    }
+}
+
+extension FxAPushMessageHandler {
+    func handleCollectionChanged(_ data: JSON?) -> Success {
+        guard let collections = data?["collections"].arrayObject as? [String] else {
+            log.warning("collections_changed received but incomplete: \(data)")
+            return deferMaybe(PushMessageError.messageIncomplete)
+        }
+        let sm = profile.syncManager
+        let deferreds = collections.flatMap { (id: String) -> SyncResult? in
+            switch id {
+            case "addons":
+                return nil
+            case "forms":
+                return nil
+            case "prefs":
+                return nil
+            case "bookmarks":
+                return nil
+            case "history":
+                return sm.syncHistory()
+            case "tabs":
+                return sm.syncClientsThenTabs()
+            case "passwords":
+                return sm.syncLogins()
+            case "clients":
+                // if we get "tabs", "clients" then we'll do "tabs" first, 
+                // then "client", but not "tabs" a second time.
+                return sm.syncClients()
+            default:
+                return nil
+            }
+        }
+
+        if deferreds.isEmpty {
+            return deferMaybe(PushMessageError.noActionNeeded)
+        }
+
+        return all(deferreds).bind { _ in
```

Comment:
Dang, I had considered exposing a `syncNamedCollections`, but rejected it because I didn't want to have `SyncManager` dealing with aliasing `passwords` -> `logins`.

I've added the suggested method to `syncManager`, and a `SyncReason.push`. Suggest a re-review. /cc @sleroux